apps/heft/src/utilities/GitUtilities.ts

```diff
@@ -109,8 +112,68 @@ export class GitUtilities {
     }
   }
 
+  public async getUnignoredFilesAsync(): Promise<Set<string>> {
+    this._ensureGitMinimumVersion({ major: 2, minor: 22, patch: 0 });
+    this._ensurePathIsUnderGitWorkingTree();
+
+    const result: IExecuteGitCommandResult = await this._executeGitCommandAndCaptureOutputAsync({
+      command: 'ls-files',
+      args: ['--cached', '--modified', '--others', '--deduplicate', '--exclude-standard']
+    });
+
+    if (result.exitCode !== 0) {
+      throw new Error(
+        `The "git ls-files" command failed with status ${result.exitCode}: ` + result.errorLines.join('\n')
+      );
+    }
+
+    // Return the set of unignored files. The output is relative to the working directory, so join
+    // them to make an absolute path.
+    return new Set(result.outputLines.map((line: string) => path.join(this._workingDirectory, line)));
+  }
+
+  public async checkIgnoreAsync(filePaths: Iterable<string>): Promise<Set<string>> {
+    this._ensurePathIsUnderGitWorkingTree();
+
+    const gitInfo: IGitRepoInfo = this.getGitInfo()!;
+    const gitRepoRootPath: string = gitInfo.root;
+    const ignoreMatcherMap: Map<string, IIgnoreMatcher> = await this._getIgnoreMatchersAsync(gitRepoRootPath);
+    const ignoredFiles: Set<string> = new Set();
+
+    for (const filePath of filePaths) {
+      if (!path.isAbsolute(filePath)) {
+        throw new Error(`The filePath must be an absolute path: "${filePath}"`);
+      }
+
+      // Find the best matcher for the file path by finding the longest matcher path that is a prefix to
+      // the file path
+      let longestMatcherPath: string | undefined;
+      let foundMatcher: IIgnoreMatcher | undefined;
+      for (const [matcherPath, matcher] of ignoreMatcherMap) {
+        if (filePath.startsWith(matcherPath) && matcherPath.length > (longestMatcherPath?.length || 0)) {
```

Comment:
Talked offline, while I don't disagree, going to avoid for 3 reasons:
- It's currently in rush-lib and that would complicate publishing release versions
- It doesn't support leaf node traversal yet
- The `startsWith` check checks against every ignorer, but the ignorers are only created per-gitignore under a project. Because there are generally 0-1 .gitignores in any given project (with a few exceptions), the general case is just not affected all that much by doing a `startsWith` check for all ignorers

Will add a comment